### PR TITLE
odo dev --debug

### LINKED ISSUE #5583: odo dev --debug - debugging components
## User Story
- As an odo user
- I want to be able to debug my application during the development
- So that I can debug it :smile: 

## Acceptance Criteria
- [ ] when the user runs `odo dev` with `--debug` code it should execute the default `kind: debug` command as defined in `devfile.yaml` instead of `kind: run` command.
- [ ] if there is no debug command defined in devfile it should error out with a message explaining why the debug can't be executed
- [ ] tests
- [ ] documentation

/kind user-story

### PR DESCRIPTION
**Pull request by feloy**
**What type of PR is this:**

/kind feature

**What does this PR do / why we need it:**

**Which issue(s) this PR fixes:**

Fixes #5583 

**PR acceptance criteria:**

- [ ] Unit test 

- [x] Integration test 

- [ ] Documentation 

**How to test changes / Special notes to the reviewer:**

### CODE REVIEW

File: tests/integration/devfile/cmd_dev_debug_test.go [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+				devSession.WaitEnd()
+			})
+			It("should expect a ws connection when tried to connect on default debug port locally", func() {
+				// 400 response expected because the endpoint expects a websocket request and we are doing a HTTP GET
+				// We are just using this to validate if nodejs agent is listening on the other side
+				helper.HttpWaitForWithStatus("http://"+ports["5858"], "WebSockets request was expected", 12, 5, 400)
+			})
+		})
+	})
+
```
Comment (rm3l) [inline tests/integration/devfile/cmd_dev_debug_test.go:56]: Would it make sense to also add a test case for Devfiles that do not define any debug command? The behavior from what I've seen in this PR is to make `odo` exit with an error message.

**Review by rm3l — CHANGES_REQUESTED** [review]

**Review by rm3l — APPROVED** [review]

**Review by dharmit — CHANGES_REQUESTED** [review]

**Comment by dharmit** [comment]
@feloy in the past when we were working on debug mode, we used to consider testing things using VS Code `launch.json` and connecting to the port on which the app's debug mode port is listening.

For this PR, I downloaded the devfile in the tests of this PR, but can't really connect to it. Or, maybe, VS Code is connecting to it, but I don't know what to do. Also, I have never worked with websockets myself, so maybe I'm doing it completely wrong.

Do you think you can share the steps you followed to test the debug mode yourself?

I spent some time on it, but still feel dumb about this thing (how to test debug mode functionality.)

**Comment by feloy** [comment]
> @feloy in the past when we were working on debug mode, we used to consider testing things using VS Code `launch.json` and connecting to the port on which the app's debug mode port is listening.
> 
> For this PR, I downloaded the devfile in the tests of this PR, but can't really connect to it. Or, maybe, VS Code is connecting to it, but I don't know what to do. Also, I have never worked with websockets myself, so maybe I'm doing it completely wrong.
> 
> Do you think you can share the steps you followed to test the debug mode yourself?
> 
> I spent some time on it, but still feel dumb about this thing (how to test debug mode functionality.)

@dharmit It should be the exact same way as with odo v2. If you follow the steps described in this blog post (Debugging the micro-service), it should work: https://dev.to/feloy/odo-a-tool-to-simplify-development-on-kubernetes-5c9a

To test with the files used in the integration test, you can get the launch.json from the blog post, then run a Debug session and set a Breakpoint where a request is replied (on the `res.send('Hello from Node.js Starter Application!');` for example). Then you execute a request on your app endpoint (`curl 127.0.0.1:40001`) and vscode should pop up and show your program interrupted on your breakpoint

**Comment by dharmit** [comment]
> To test with the files used in the integration test, you can get the launch.json from the blog post, then run a Debug session and set a Breakpoint where a request is replied (on the `res.send('Hello from Node.js Starter Application!');` for example). Then you execute a request on your app endpoint (`curl 127.0.0.1:40001`) and vscode should pop up and show your program interrupted on your breakpoint

@feloy I am doing something wrong, it seems.

`launch.json`:
```json
{
    "version": "0.2.0",
    "configurations": [
        {
            "type": "node",
            "request": "attach",
            "name": "odo dev debug",
            "program": "${workspaceFolder}/server.js",
            "address": "localhost",
            "port": 40002,
            "localRoot": "${workspaceFolder}",
        }
    ]
}
```
I have added breakpoints at both the lines in the `server.js` that have `Hello from Node.js ` in them. I run `odo dev --debug`, then I start debugging session from VS Code, and finally do `curl localhost:40001`/`curl localhost:40002`. Neither of them takes me to the breakpoint.

The feeling of dumbness is getting real bad at this point, so I'll approve since I don't want my limitation to block your PR.

/approve
/lgtm

**Comment by feloy** [comment]
> > To test with the files used in the integration test, you can get the launch.json from the blog post, then run a Debug session and set a Breakpoint where a request is replied (on the `res.send('Hello from Node.js Starter Application!');` for example). Then you execute a request on your app endpoint (`curl 127.0.0.1:40001`) and vscode should pop up and show your program interrupted on your breakpoint
> 
> @feloy I am doing something wrong, it seems.
> 
> `launch.json`:
> 
> ```json
> {
>     "version": "0.2.0",
>     "configurations": [
>         {
>             "type": "node",
>             "request": "attach",
>             "name": "odo dev debug",
>             "program": "${workspaceFolder}/server.js",
>             "address": "localhost",
>             "port": 40002,
>             "localRoot": "${workspaceFolder}",
>         }
>     ]
> }
> ```
> 
> I have added breakpoints at both the lines in the `server.js` that have `Hello from Node.js ` in them. I run `odo dev --debug`, then I start debugging session from VS Code, and finally do `curl localhost:40001`/`curl localhost:40002`. Neither of them takes me to the breakpoint.
> 
> The feeling of dumbness is getting real bad at this point, so I'll approve since I don't want my limitation to block your PR.
> 
> /approve /lgtm

Can you try with this launch.json (remoteRoot added):

```
{
    "version": "0.2.0",
    "configurations": [
        {
            "name": "Attach to remote 5858",
            "type": "node",
            "request": "attach",
            "address": "localhost",
            "port": 40002,
            "localRoot": "${workspaceFolder}",
            "remoteRoot": "/projects"
        }
    ]
}
```

**Comment by feloy** [comment]
/override ci/prow/v4.10-integration-e2e

IBM Could tests pass

**Comment by dharmit** [comment]
> Can you try with this launch.json:

That works. Thanks!

It works when I curl the port 40001, not 40002. However, the debug port 5858 is forwarded to 40002:
```
Your application is now running on the cluster
 - Forwarding from 127.0.0.1:40001 -> 3000
 - Forwarding from 127.0.0.1:40002 -> 5858
```
I think I need to use websockets instead of http when trying to work with 40002, right?

**Comment by feloy** [comment]
Port 5858 (and 40002) are for the debugger only, for it to be able to communicate with the process inside the container. You, as a developer, don't need to use it directly.

**Comment by dharmit** [comment]
OK, thanks for guiding me patiently. I surely need some debugging 101 class/course. 😞

**Pull request merged**